lib/Frontend/nvqpp/ConvertExpr.cpp

```diff
@@ -1985,34 +1974,65 @@ bool QuakeBridgeVisitor::VisitCXXConstructExpr(clang::CXXConstructExpr *x) {
+      // second is the allocator
+      if (ctorName == "vector" && x->getNumArgs() == 2) {
+        // Here we create a lambda that will peel off the sugar types
+        auto peelOffClangTypes = [&](clang::QualType type) {
+          clang::QualType lastType = type;
+          while (true) {
+            type = type.getSingleStepDesugaredType(*astContext);
+            if (type == lastType)
+              break;
+            lastType = type;
+          }
+
+          return lastType;
+        };
+
         // This is a std::vector constructor, first we'll check if it
         // is constructed from a constant initializer list, in that case
         // we'll have a AllocaOp at the top of the stack that allocates a
         // ptr<array<TxC>>, where C is constant / known
-        if (auto alloca = dyn_cast<cc::AllocaOp>(peekValue().getDefiningOp())) {
-          auto asPtrType =
-              dyn_cast<cc::PointerType>(alloca.getAddress().getType());
-          if (auto arrayTy =
-                  dyn_cast<cc::ArrayType>(asPtrType.getElementType()))
-            if (alloca.getNumOperands() > 0)
-              return pushValue(builder.create<cc::StdvecInitOp>(
-                  loc, cc::StdvecType::get(arrayTy.getElementType()), alloca,
-                  alloca.getSeqSize()));
+        auto type = x->getArg(0)->getType();
+        auto desugared = peelOffClangTypes(type);
+        if (auto recordType =
+                dyn_cast<clang::RecordType>(desugared.getTypePtr())) {
+          if (recordType->getDecl()->getName().equals("initializer_list")) {
+            auto allocation = popValue();
+            if (auto ptrTy = dyn_cast<cc::PointerType>(allocation.getType())) {
+              if (auto arrayTy =
+                      dyn_cast<cc::ArrayType>(ptrTy.getElementType())) {
+                if (auto definingOp = allocation.getDefiningOp<cc::AllocaOp>())
+                  return pushValue(builder.create<cc::StdvecInitOp>(
+                      loc, cc::StdvecType::get(arrayTy.getElementType()),
+                      allocation, definingOp.getSeqSize()));
+              }
+            }
+          }
         }
 
         // Next check if its created from a size integer
-        if (peekValue().getType().isIntOrFloat()) {
-          auto arrSize = popValue();
-          auto dataType = popType();
-
-          // create stdvec init op without a buffer.
-          // Allocate the required memory chunk
-          Value alloca = builder.create<cc::AllocaOp>(loc, dataType, arrSize);
-
-          // Create the stdvec_init op
-          return pushValue(builder.create<cc::StdvecInitOp>(
-              loc, cc::StdvecType::get(dataType), alloca, arrSize));
+        // Let's do a check on the first argument, make sure that when
+        // we peel off all the typedefs that it is an integer
+        if (auto builtInType =
+                dyn_cast<clang::BuiltinType>(desugared.getTypePtr())) {
+          if (builtInType->isInteger() &&
```

Comment:
Do we want to add `|| isFloat()`?

Reply:
For `std::vector<T>(std::size_t)` I would have thought that at this point in the AST it would not be possible for this type to be a float, that it had to be an integer. Someone could have programmed a float as input, but there would already be an implicit cast to an integer at this point.